src/coreComponents/physicsSolvers/fluidFlow/wells/CompositionalMultiphaseWell.cpp

```diff
@@ -235,6 +235,23 @@ void CompositionalMultiphaseWell::registerDataOnMesh( Group & meshBodies )
         reference().resizeDimension< 0 >( m_numComponents );
       wellControls.registerWrapper< real64 >( viewKeyStruct::dCurrentTotalVolRate_dRateString() ).
         setRestartFlags( RestartFlags::NO_WRITE );
+
+      // write rates output header
+      if( getLogLevel() > 0 )
+      {
+        string const wellControlsName = wellControls.getName();
+        string const massUnit = m_useMass ? "kg" : "mol";
+        integer const useSurfaceConditions = wellControls.useSurfaceConditions();
+        string const conditionKey = useSurfaceConditions ? "surface" : "reservoir";
+        string const unitKey = useSurfaceConditions ? "s" : "r";
+        integer const numPhase = m_numPhases;
+        // format: time,bhp,total_rate,total_vol_rate,phase0_vol_rate,phase1_vol_rate,...
+        std::ofstream outputFile( m_ratesOutputDir + "/" + wellControlsName + ".csv" );
+        outputFile << "time [s],bhp [Pa],total rate ["<<massUnit<<"/s],total " << conditionKey << " volumetric rate ["<<unitKey<<"m3/s]";
```

Comment:
Is this thread safe ?

Reply:
my hope is the same - we are writing here separate file for each `wellControls`, so hopefully it's safe
would appreciate if @klevzoff can have a look and correct me if i am wrong

Reply:
We don't parallelize across multiple wells (at least not yet), so it would even be safe to write to the same file

Reply:
Looking at another comment by @jafranc, I guess this was about MPI/multiprocess safety. Yeah, you want only one rank to write data for each well. Regardless of how the well is partitioned across ranks, each rank has each well object (even if it has no locally present segments). Only one rank will be considered the nominal "owner" of the well and be responsible for assembling the control well equation, and it's usually the one that owns the well head IIRC. This is probably the rank you want writing well rates. This is determined via `subRegion.isLocallyOwned()`, and `printRates()` already checks that condition. You just need to check it here when writing the file header too.

Reply:
added the check, that you